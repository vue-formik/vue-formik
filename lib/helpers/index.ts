/* eslint-disable @typescript-eslint/no-explicit-any */
const getNestedValue = <T extends object>(obj: T, path: string) => {
  if (!obj || typeof obj !== "object") return undefined;

  if (path in obj) {
    return obj[path as keyof T];
  }

  const keys = path.split('.'); // Split the path into keys

  if (keys.length === 1) {
    if (/^.*\[\d+]$/.test(keys[0])) {
      const [key, index] = keys[0].split("[");
      const i = parseInt(index.replace("]", ""), 10);
      const element = obj[key as keyof T];
      if (Array.isArray(element)) {
        return element[i];
      } else {
        return undefined;
      }
    }
    return obj[keys[0] as keyof T];
  }

  return getNestedValue(obj[keys[0] as keyof T] as T, keys.slice(1).join('.')); // Recursively get the nested value
};

const updateNestedProperty = <T extends Record<string, any>>(
  obj: T,
  path: string,
  value: any,
): void => {
  if (!obj || typeof obj !== "object") {
    obj = {} as T;
    obj[path as keyof T] = value;
    return;
  }

  if (path in obj) {
    // update object here
    obj[path as keyof T] = value;
    return;
  }

  const keys = path.split('.');

  if (keys.length === 1) {
    if (/^.*\[\d+]$/.test(keys[0])) {
      const [key, index] = keys[0].split("[");
      if (!Array.isArray(obj[key as keyof T])) {
        obj[key as keyof T] = [];
      }
      const i = parseInt(index.replace("]", ""), 10);

      obj[key as keyof T][i] = value;
      return;
    }
    obj[keys[0] as keyof T] = value;
    return;
  }
  // Split the path into keys
  return updateNestedProperty(getNestedValue(obj, keys[0]), keys.slice(1).join('.'), value); // Recursively get the nested value
};

function clearReactiveObject<T extends object>(obj: T): void {
  if (typeof obj !== "object" || obj === null) {
    return; // Skip non-object or null values
  }

  // Skip unsupported types like Map, Set, WeakMap, WeakSet
  if (
    obj instanceof Map ||
    obj instanceof WeakMap ||
    obj instanceof Set ||
    obj instanceof WeakSet
  ) {
    return;
  }

  // Check for Vue's reactive types and unwrap them if needed
  if ("__v_isRef" in obj || "__v_isReactive" in obj) {
    return;
  }

  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      const value = obj[key as keyof T];
      if (value && typeof value === "object" && !Array.isArray(value)) {
        clearReactiveObject(value); // Recursively clear nested objects
      }
      delete obj[key as keyof T]; // Delete the property
    }
  }
}

export { updateNestedProperty, getNestedValue, clearReactiveObject };
