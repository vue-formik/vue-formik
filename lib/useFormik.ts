import { computed, reactive, toRaw, watch } from 'vue'
import { ObjectSchema } from 'yup'
import type { FormikHelpers, IUseFormik, ValidationRule } from '@/types'
import { getNestedValue, updateNestedProperty } from '@/helpers'

const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
}: IUseFormik<T>) => {
  const validate = (values: T): Partial<Record<keyof T, string>> => {
    const validationErrors: Partial<Record<keyof T, string>> = {};

    if (validationSchema instanceof ObjectSchema) {
      try {
        const vSchema = validationSchema as ObjectSchema<T>;
        vSchema.validateSync(values, { abortEarly: false });
      } catch (e) {
        const err = e as never as { inner: { path: string; message: string }[] };
        if (typeof err === "object" && Array.isArray(err?.inner)) {
          err.inner.forEach((error) => {
            validationErrors[error.path as keyof T] = error.message;
          });
        }
      }
    } else {
      for (const key in validationSchema) {
        const value = values[key as keyof T];
        const rule = validationSchema[key as keyof T] as ValidationRule<T[keyof T]>;

        if (typeof rule === 'function') {
          const error = rule(value);
          if (error) {
            validationErrors[key as keyof T] = error;
          }
        }
      }
    }

    return validationErrors;
  };

  const values = reactive(initialValues);
  const errors = reactive(validate(initialValues));
  const touched = reactive({} as Partial<Record<keyof T, boolean | Record<string, boolean>>>);

  const setValues = (newValues: T) => {
    Object.assign(values, newValues); // Safely copy values into reactive object
    Object.assign(errors, validate(newValues)); // Validate updated values
  };

  const setErrors = (newErrors: T) => {
    Object.assign(errors, newErrors);
  };

  const reset = () => {
    Object.assign(values, initialValues);
    Object.assign(errors, validate(initialValues));
  };

  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, touchedValue: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, touchedValue);
  }

  const handleSubmit = (e: Event) => {
    e.preventDefault();
    onSubmit(
      toRaw(values) as T,
      {
        setErrors,
        reset,
        setValues,
      } as FormikHelpers<T>
    );
  };

  const handleBlur = (e: FocusEvent) => {
    const fieldName = (e.target as HTMLInputElement).name as keyof T;
    setFieldTouched(fieldName.toString(), true);
  }
  const handleChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const fieldName = target.name as keyof T;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    setFieldValue(fieldName.toString(), value);
  }

  const isValid = computed(() => {
    return Object.keys(errors).length === 0;
  })

  const isDirty = computed(() => {
    return JSON.stringify(values) !== JSON.stringify(initialValues);
  })

  watch(
    () => values,
    () => {
      const validationErrors = validate(toRaw(values) as T);
      //@ts-expect-error Object.keys(errors) is not iterable
      Object.keys(errors).forEach((key) => delete errors[key]); // Clear existing errors
      Object.assign(errors, validationErrors); // Assign new errors
    },
    { deep: true }
  );


  const hasFieldError = (field: string) => {
    const errorValue = getNestedValue(errors, field);
    const touchedValue = getNestedValue(touched, field);
    return (errorValue !== undefined && !!touchedValue);
  }

  const getFieldError = (field: string) => {
    if (hasFieldError(field)) {
      return errors[field as keyof typeof errors];
    } else {
      return '';
    }
  }

  return {
    values,
    errors,
    touched,
    setValues,
    reset,
    setFieldValue,
    setFieldTouched,
    handleBlur,
    handleChange,
    handleSubmit,
    hasFieldError,
    getFieldError,
    isValid,
    isDirty
  };
};

export default useFormik;
