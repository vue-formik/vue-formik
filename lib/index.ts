import useFormik from "@/composables/useFormik";
import FormTextArea from "@/components/FormTextArea.vue";
import FormTextField from "@/components/FormTextField.vue";

export { FormTextArea, FormTextField, useFormik };
