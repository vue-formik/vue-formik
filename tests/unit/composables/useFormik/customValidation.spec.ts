import { test, expect, describe } from "vitest";
import { useFormik } from "../../../../lib";
import { nextTick } from "vue";

const validateRequiredField = (value: string | undefined) =>
  value ? undefined : "This field is required";

const validateMinLength = (value: string, minLength: number) =>
  value.length >= minLength ? undefined : `Must be at least ${minLength} characters`;

const validateArrayField = (values: string[], minLength: number) =>
  values.map((value) => validateRequiredField(value) || validateMinLength(value, minLength));

const validateObjectField = (
  value: { [key: string]: string },
  rules: { [key: string]: (val: string) => string | undefined },
) => {
  const errors: { [key: string]: string } = {};
  Object.keys(rules).forEach((key) => {
    const error = rules[key](value[key]);
    if (error) errors[key] = error;
  });
  return Object.keys(errors).length ? errors : undefined;
};

describe("useFormik custom validation", async () => {
  test("should validate form with custom validation schema", async () => {
    const initialValues = { contact: "" };
    const validationSchema = {
      contact: (value: string) => validateRequiredField(value),
    };
    const { errors } = useFormik({ initialValues, validationSchema });
    expect(errors).toMatchSnapshot();
  });

  test("should validate array field with custom validation schema", async () => {
    const initialValues = { contacts: ["", ""] };
    const validationSchema = {
      contacts: (values: string[]) => validateArrayField(values, 3),
    };
    const { errors } = useFormik({ initialValues, validationSchema });
    expect(errors).toMatchSnapshot();
  });

  test("should validate object field with custom validation schema", async () => {
    const initialValues = { address: { street: "", city: "" } };
    const validationSchema = {
      address: (value: { street: string; city: string }) =>
        validateObjectField(value, {
          street: (val) => validateRequiredField(val) || validateMinLength(val, 3),
          city: validateRequiredField,
        }),
    };
    const { errors } = useFormik({ initialValues, validationSchema });
    await nextTick();
    expect(errors).toMatchSnapshot();
  });

  describe("validation on value update", () => {
    test("should validate form on value update with custom validation schema", async () => {
      const initialValues = { contact: "" };
      const validationSchema = {
        contact: (value: string) => validateRequiredField(value) || validateMinLength(value, 3),
      };
      const { errors, setValues } = useFormik({ initialValues, validationSchema });
      expect(errors).toMatchSnapshot();

      setValues({ contact: "1" });
      await nextTick();
      expect(errors).toMatchSnapshot();

      setValues({ contact: "123" });
      await nextTick();
      expect(errors).toMatchSnapshot();
    });

    test("should validate array field on value update with custom validation schema", async () => {
      const initialValues = { contacts: ["", ""] };
      const validationSchema = {
        contacts: (values: string[]) => validateArrayField(values, 3),
      };
      const { errors, setValues } = useFormik({ initialValues, validationSchema });
      expect(errors).toMatchSnapshot();

      setValues({ contacts: ["1", "12"] });
      await nextTick();
      expect(errors).toMatchSnapshot();

      setValues({ contacts: ["123", "123"] });
      await nextTick();
      expect(errors).toMatchSnapshot();
    });

    test("should validate object field on value update with custom validation schema", async () => {
      const initialValues = { address: { street: "", city: "" } };
      const validationSchema = {
        address: (value: { street: string; city: string }) =>
          validateObjectField(value, {
            street: (val) => validateRequiredField(val) || validateMinLength(val, 3),
            city: validateRequiredField,
          }),
      };
      const { errors, setValues } = useFormik({
        initialValues,
        validationSchema,
      });
      await nextTick();
      expect(errors).toMatchSnapshot();

      setValues({ address: { street: "1", city: "" } });
      await nextTick();
      expect(errors).toMatchSnapshot();

      setValues({ address: { street: "123", city: "1" } });
      await nextTick();
      expect(errors).toMatchSnapshot();
    });
  });

  describe("validation on blur", () => {
    test("should validate form on blur with custom validation schema", async () => {
      const initialValues = { contact: "" };
      const validationSchema = {
        contact: (value: string) => validateRequiredField(value) || validateMinLength(value, 3),
      };
      const { errors, setTouched } = useFormik({ initialValues, validationSchema });
      expect(errors).toMatchSnapshot();

      setTouched({ contact: true });
      expect(errors).toMatchSnapshot();
    });

    test("should validate array field on blur with custom validation schema", async () => {
      const initialValues = { contacts: ["", ""] };
      const validationSchema = {
        contacts: (values: string[]) => validateArrayField(values, 3),
      };
      const { errors, setTouched } = useFormik({ initialValues, validationSchema });
      expect(errors).toMatchSnapshot();

      setTouched({ contacts: [true, true] });
      expect(errors).toMatchSnapshot();
    });

    test("should validate object field on blur with custom validation schema", async () => {
      const initialValues = { address: { street: "", city: "" } };
      const validationSchema = {
        address: (value: { street: string; city: string }) =>
          validateObjectField(value, {
            street: (val) => validateRequiredField(val) || validateMinLength(val, 3),
            city: validateRequiredField,
          }),
      };
      const { errors, setTouched } = useFormik({ initialValues, validationSchema });
      expect(errors).toMatchSnapshot();

      setTouched({ address: { street: true, city: true } });
      expect(errors).toMatchSnapshot();
    });
  });

  test("custom validation with nested objects", () => {
    const initialValues = {
      address: {
        street: "",
        city: "",
      },
    };

    const validationSchema = {
      address: {
        street: (value: string) => validateRequiredField(value),
        city: (value: string) => validateRequiredField(value),
      },
    };

    const { errors } = useFormik({ initialValues, validationSchema });
    expect(errors).toMatchSnapshot();
  });

  test("custom validation with a function", () => {
    const initialValues = {
      name: "",
      email: "",
    };

    const validationSchema = (values: typeof initialValues) => {
      const errors: Partial<typeof initialValues> = {};
      if (!values.name) {
        errors.name = "Name is required";
      }
      if (!values.email) {
        errors.email = "Email is required";
      }
      return errors;
    };

    const { errors } = useFormik({
      initialValues,
      validationSchema,
    });
    expect(errors).toMatchSnapshot();
  });

  test("dot notation for nested fields", () => {
    const initialValues = {
      address: {
        street: "",
        city: "",
      },
    };

    const validationSchema = {
      "address.street": (value: string) => validateRequiredField(value),
      "address.city": (value: string) => validateRequiredField(value),
    };

    const { errors } = useFormik({ initialValues, validationSchema });
    expect(errors).toMatchSnapshot();
  });
});
