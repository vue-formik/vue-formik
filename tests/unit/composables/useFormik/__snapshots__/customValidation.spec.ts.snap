// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFormik custom validation > should validate array field with custom validation schema 1`] = `
{
  "contacts": [
    "This field is required",
    "This field is required",
  ],
}
`;

exports[`useFormik custom validation > should validate form with custom validation schema 1`] = `
{
  "contact": "This field is required",
}
`;

exports[`useFormik custom validation > should validate form with custom validation schema 2`] = `""`;

exports[`useFormik custom validation > should validate object field with custom validation schema 1`] = `
{
  "address": {
    "city": "This field is required",
    "street": "This field is required",
  },
}
`;

exports[`useFormik custom validation > validation on blur > should validate array field on blur with custom validation schema 1`] = `
{
  "contacts": [
    "This field is required",
    "This field is required",
  ],
}
`;

exports[`useFormik custom validation > validation on blur > should validate array field on blur with custom validation schema 2`] = `
{
  "contacts": [
    "This field is required",
    "This field is required",
  ],
}
`;

exports[`useFormik custom validation > validation on blur > should validate form on blur with custom validation schema 1`] = `
{
  "contact": "This field is required",
}
`;

exports[`useFormik custom validation > validation on blur > should validate form on blur with custom validation schema 2`] = `
{
  "contact": "This field is required",
}
`;

exports[`useFormik custom validation > validation on blur > should validate object field on blur with custom validation schema 1`] = `
{
  "address": {
    "city": "This field is required",
    "street": "This field is required",
  },
}
`;

exports[`useFormik custom validation > validation on blur > should validate object field on blur with custom validation schema 2`] = `
{
  "address": {
    "city": "This field is required",
    "street": "This field is required",
  },
}
`;

exports[`useFormik custom validation > validation on value update > should validate array field on value update with custom validation schema 1`] = `
{
  "contacts": [
    "This field is required",
    "This field is required",
  ],
}
`;

exports[`useFormik custom validation > validation on value update > should validate array field on value update with custom validation schema 2`] = `
{
  "contacts": [
    "Must be at least 3 characters",
    "Must be at least 3 characters",
  ],
}
`;

exports[`useFormik custom validation > validation on value update > should validate array field on value update with custom validation schema 3`] = `
{
  "contacts": [
    undefined,
    undefined,
  ],
}
`;

exports[`useFormik custom validation > validation on value update > should validate form on value update with custom validation schema 1`] = `
{
  "contact": "This field is required",
}
`;

exports[`useFormik custom validation > validation on value update > should validate form on value update with custom validation schema 2`] = `
{
  "contact": "Must be at least 3 characters",
}
`;

exports[`useFormik custom validation > validation on value update > should validate form on value update with custom validation schema 3`] = `{}`;

exports[`useFormik custom validation > validation on value update > should validate object field on value update with custom validation schema 1`] = `
{
  "address": {
    "city": "This field is required",
    "street": "This field is required",
  },
}
`;

exports[`useFormik custom validation > validation on value update > should validate object field on value update with custom validation schema 2`] = `
{
  "address": {
    "city": "This field is required",
    "street": "Must be at least 3 characters",
  },
}
`;

exports[`useFormik custom validation > validation on value update > should validate object field on value update with custom validation schema 3`] = `{}`;
